crates/sui-types/src/temporary_store.rs

```diff
@@ -772,6 +772,47 @@ impl<S> TemporaryStore<S> {
     }
 }
 
+impl<S: GetModule> TemporaryStore<S> {
+    /// Check that this transaction neither creates nor destroys SUI. This should hold for all txes except
+    /// the epoch change tx, which mints staking rewards equal to the gas fees burned in the previous epoch.
+    /// This intended to be called *after* we have charged for gas + applied the storage rebate to the gas object
+    pub fn check_sui_conserved(&self) {
+        // total SUI in input objects
+        let input_sui = self.mutable_input_refs.iter().fold(0, |acc, o| {
+            acc + self
```

Comment:
This doesn't seem to include deleted objects (`self.deleted`), which I suppose are objects that are passed in by value? The sui stored in the deleted objects should be part of the input sui too.

Reply:
`mutable_input_refs` includes both owned and `&mut` inputs.

However, you made me realize that I forgot about dynamic fields, which will definitely break things... let me take a look at that.

Reply:
Right. I think `written` should already include modified and new dynamic fields but you will need to check `deleted` for deleted dynamic fields.

Reply:
For now I am just skipping this check if there are any dynamic fields + asking for help on the "before" lookup.

Reply:
Good point re: deleted--I will fix that